### DISCUSSION TITLE
fetch gives a very opaque error when failing

### DISCUSSION THREAD

**Discussion by vedantroy** [vercel/next.js#53931] [Help]
### Summary

Was wondering if anyone else has noticed this. Sometimes I get really opaque errors of the form "fetch failed" with no stacktrace, it looks like the error is originating from [undici](https://github.com/nodejs/undici).

The thing that is confusing is--this looks like it is happening inside of client components in the app directory, and I assumed client components are using the browser's native fetch. So not sure what is going on here.

Anyone ran into this problem & what do y'all do about it?

**Comment by devjiwonchoi** [comment]
Assuming you are using env variables for the API endpoint, in order to use env variables on Client Components, you need to expose it to the client by naming the variable starting with `NEXT_PUBLIC_`.

If this is not the case, please provide reproduction or codes.

### Reference

- [Bundling Environment Variables for the Browser](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser)

**Comment by icyJoseph** [comment]
Got any screenshots or code snippets? Maybe you are taking something for granted?

> this looks like it is happening inside of client components in the app directory, and I assumed client components are using the browser's native fetch.

That is only partially correct. During SSR those client components use the Node available fetch. However, since in client components, we often fetch inside use-effects, or rather, `render then fetch`, it should be using the browser available fetch.

As said above, without code snippets, it is really hard to guess anything. Undici is the "engine" behind the Node Native fetch, so it ought to be something server side, but what?
